### Dependencies

Using `git` and `mamba` and `pip`:

```
git clone https://github.com/nspope/singer-snakemake my-singer-run && cd my-singer-run
mamba install -c bioconda snakemake
python3 -m pip install -r requirements.txt
snakemake --cores=20 --configfile=configs/example_config.yaml
```

### Inputs

The input files for each chromosome are:

  - *chromosome_name.vcf* VCF that can be used as SINGER input (diploid, phased, not compressed)
  - *chromosome_name.mask.bed* (optional) bed file containing inaccessible intervals
  - *chromosome_name.hapmap* (optional) recombination map in the format described in the documentation for `msprime.RateMap.read_hapmap` (see [here](https://tskit.dev/msprime/docs/stable/api.html#msprime.RateMap.read_hapmap))

see `example/*`.

### Config

A template for the configuration file is in `configs/example_config.yaml`:

```
...
```

### Outputs

The output files for each chromosome will be generated in `results/chromosome_name`:

  - _chromosome_name.adjusted_mu.p_ : _msprime.RateMap_ containing adjusted mutation rates (`proportion_accessible_bases * mutation_rate`) in each chunk
  - _chromosome_name.replicate.trees_ : a tree sequence replicate generated by SINGER
  - _chromosome_name.replicate.stats.p_ : "fitted values" for summary statistics (e.g. branch-mode statistics calculated with tskit)
  - _chromosome_name.stats.p_ : "observed values" for summary statistics (e.g. calculated from with `scikit-allel`)
  - _chunks/*_ the raw SINGER output and logs
  - _diversity-trace.png_, _tajima-d-trace.png_ : MCMC trace for fitted nucleotide diversity and Tajima's D
  - _diversity-scatter.png_, _tajima-d-scatter.png_ : Observed vs fitted nucleotide diversity and Tajima's D, across chunks
  - _diversity-skyline.png_, _tajima-d-skyline.png_ : Observed and fitted nucleotide diversity and Tajima's D, across genome position
  - _folded-afs.png_, _unfolded-afs.png_ : Observed vs fitted site frequency spectra
  - _site-density.png_ : Sanity check that shows proportion missing data, proportion variant bases (out of accessible bases), recombination rate across genome position.
